src/rich-log/domain.hh

```diff
@@ -14,49 +18,62 @@ namespace verbosity
 ///   - everything else is enabled
 enum type : int
 {
+    _None = 0,
+
     /// TRACE is used to provide diagnostic information verbosely and in very small steps
     /// e.g. logging in each iteration of an algorithm
-    Trace = 0,
+    Trace,
 
     /// DEBUG is used to provide diagnostic information that usually do not matter
     /// e.g. logging the arguments to an algorithm
-    Debug = 1,
+    Debug,
 
     /// INFO is used to log interesting information that USERS of your algorithm might want to know about
     /// e.g. the selected graphics card, filenames for large save operations, UI actions
-    Info = 2,
+    Info,
 
     /// WARNING is used to log suspicious calls / inputs
     /// e.g. loading empty files, showing empty meshes, degenerate faces
     /// while they are not strictly errors, you should minimize false positives (users will start to ignore them)
-    Warning = 3,
+    Warning,
 
     /// ERRROR is used for clear errors and wrong use of APIs
     /// e.g. trying to load a nonexistant file, providing unsupported argument combinations
     /// these errors do not crash the program, but are clearly wrong
-    Error = 4,
+    Error,
 
     /// FATAL is used for errors that will, with a high probability, crash/invalidate the program
     /// e.g. failure to acquire crucial resources, out of memory
     /// fatal errors also tend to be unrecoverable with no clear way to fix them
-    Fatal = 5,
+    Fatal,
 
-    _count
+    _EndRange,
+    _NumVerbosities = _EndRange - 1
 };
 }
 
-struct domain_info
+struct RLOG_API domain_info
 {
-    rlog::verbosity::type min_verbosity = rlog::verbosity::Info; // always first
-    char const* name = "";
-    char const* ansi_color_code = "\u001b[38;5;244m";
-
-    static constexpr domain_info make_named(char const* name)
-    {
-        domain_info di;
-        di.name = name;
-        return di;
-    }
+    rlog::verbosity::type min_verbosity = rlog::verbosity::_None;
+    rlog::verbosity::type const compile_time_min_verbosity = rlog::verbosity::_None;
+
+    char const* const name = nullptr;
+    char const* const description = nullptr;
+    char const* const ansi_color_code = nullptr;
+    uint64_t name_hash = 0;
+
+    domain_info(char const* name, char const* desc, char const* ansi_color, rlog::verbosity::type min_verbosity, int compile_min_verbosity);
+
+    static domain_info* get_first_domain();
+    static domain_info* find_domain(char const* name);
+
+    domain_info* get_next_domain() const { return next_domain; }
+
+private:
+    domain_info(domain_info const&) = delete;
+    domain_info(domain_info&&) = delete;
+
+    domain_info* next_domain = nullptr;
```

Comment:
is there really any tangible benefit to having this intrusive linked list? Otherwise, I prefer my previous vector solution

Reply:
I think for something as "core" as a logging system it's nice be as low-frills and stupid simple as possible when it comes to static init, pre-main logic, DLL loading and that entire domain of concerns. Quoting the Live++ guy where this idea was stolen from:

> The subtle difference between a static std::vector and the static pointers here is, that those pointers are of integral type, and will therefore be stored in either the .bss or .sdata segment, which means they will already be initialized long before any pre-main constructor or similar is called. So AddToList() does nothing more than adding the globally declared config value to a global linked-list. No matter in which .cpp file you add config values, and no matter in which order they are initialized, you will always have a linked-list of all ConfigSettingInt of your whole application, easily accessible and implemented in a standards-compliant way.
>
> No dynamic memory allocation, minimal pre-main work, no singletons, no hacks, no nothing. Win!

Reply:
the cons are that you have a linked list and not a contiguous space (if you want to iterate it each frame). but the biggest issue for me is the intrusiveness and API lock-in. I would be ok with this if none of that would be public api. so at least remove that part. having the domains interface as an iteration is fine by me.